functionary/train/train_lora.py

```diff
@@ -0,0 +1,396 @@
+
+    if lora_args.q_lora:
+        model = prepare_model_for_kbit_training(model, use_gradient_checkpointing=training_args.gradient_checkpointing)
+
+    model = get_peft_model(model, lora_config)
+
+    if training_args.gradient_checkpointing:
+        model.enable_input_require_grads()
+
+    model.config.use_cache = False
+    print_trainable_parameters(model)
+    return model
+
+
+def maybe_zero_3(param):
+    if hasattr(param, "ds_id"):
+        assert param.ds_status == ZeroParamStatus.NOT_AVAILABLE
+        with zero.GatheredParameters([param]):
+            param = param.data.detach().cpu().clone()
+    else:
+        param = param.detach().cpu().clone()
+    return param
+
+
+# Borrowed from peft.utils.get_peft_model_state_dict
+def get_peft_state_maybe_zero_3(named_params, bias):
+    if bias == "none":
+        to_return = {k: t for k, t in named_params if "lora_" in k}
+    elif bias == "all":
+        to_return = {k: t for k, t in named_params if "lora_" in k or "bias" in k}
+    elif bias == "lora_only":
+        to_return = {}
+        maybe_lora_bias = {}
+        lora_bias_names = set()
+        for k, t in named_params:
+            if "lora_" in k:
+                to_return[k] = t
+                bias_name = k.split("lora_")[0] + "bias"
+                lora_bias_names.add(bias_name)
+            elif "bias" in k:
+                maybe_lora_bias[k] = t
+        for k, t in maybe_lora_bias:
+            if bias_name in lora_bias_names:
+                to_return[bias_name] = t
+    else:
+        raise NotImplementedError
+    to_return = {k: maybe_zero_3(v) for k, v in to_return.items()}
+    return to_return
+
+
+def train():
+    argument_parser = transformers.HfArgumentParser((ModelArguments, DataArguments, TrainingArguments, LoraArguments))
+    model_args, data_args, training_args, lora_args = argument_parser.parse_args_into_dataclasses()
+
+    # Set RoPE scaling factor
+
+    print_rank0("lora args: ", lora_args)
+
+    print_rank0("training args: ", training_args)
+
+    model = load_model_with_rope_scaling(model_args, training_args, lora_args)
+    print_rank0(model)
+
+    tokenizer = initialize_tokenizer(
+        model,
+        model_args.model_name_or_path,
+        training_args.model_max_length,
+        training_args.cache_dir,
+    )
+
+    print_rank0("tokenizer.model_max_length: ", tokenizer.model_max_length)
+
+    model = prepare_model_for_training(model, training_args, lora_args)
+
+    assert data_args.train_data_path is not None, "Please provide a training data file."
+
+    with open(data_args.train_data_path, "r") as file:
+        raw_train_data = [json.loads(line) for line in file]
+    random.shuffle(raw_train_data)
+
+    if data_args.eval_data_path is not None:
+        with open(data_args.eval_data_path, "r") as file:
+            raw_eval_data = [json.loads(line) for line in file]
+        random.shuffle(raw_eval_data)
+
+    train_dataset = CustomDataset(raw_train_data[:100], tokenizer)
```

Comment:
100 ?

Reply:
Oh sorry, I only take first 100 items to make the training fast, forgot to remove it